File: ed/go/patterns/3.state.go
package main

type EngineState interface {
	Start() EngineState
	Stop() EngineState
}

type EngineTurnedOnState struct {
}

func (e EngineTurnedOnState) Start() EngineState {
	println("‼️ Engine already started!")
	return e
}

func (e EngineTurnedOnState) Stop() EngineState {
	println("✅ Engine stopped.")
	return EngineTurnedOffState{}
}

type EngineTurnedOffState struct {
}

func (e EngineTurnedOffState) Start() EngineState {
	println("✅ Engine started.")
	return EngineTurnedOnState{}
}

func (e EngineTurnedOffState) Stop() EngineState {
	println("‼️ Engine already stopped!")
	return e
}

type Car struct {
	engineState EngineState
}

func NewCar() Car {
	return Car{engineState: EngineTurnedOffState{}}
}

func (c *Car) StartEngine() {
	c.engineState = c.engineState.Start()
}

func (c *Car) StopEngine() {
	c.engineState = c.engineState.Stop()
}

func main() {
	c1 := NewCar()
	c1.StopEngine()

	c2 := NewCar()
	c2.StartEngine()
	c2.StopEngine()
}

File: ed/php/patterns/3.state.php
<?php
/**
 * State
 *
 * @category Behaviour
 */

interface EngineState
{
    public function startEngine();
    public function moveForward();
}

class EngineTurnedOnState implements EngineState
{
    public function startEngine()
    {
        echo 'Engine already started'.PHP_EOL;
    }

    public function moveForward()
    {
        echo 'Moved Car forward...'.PHP_EOL;
        return $this;
    }
}

class EngineTurnedOffState implements EngineState
{
    public function startEngine()
    {
        echo 'Started Engine.'.PHP_EOL;
        return new EngineTurnedOnState;
    }

    public function moveForward()
    {
        echo 'Have to start engine first!'.PHP_EOL;
    }
}

class Car implements EngineState
{
    protected $state;

    public function __construct()
    {
        $this->state = new EngineTurnedOffState;
    }

    public function startEngine()
    {
        $this->state = $this->state->startEngine();
    }

    public function moveForward()
    {
        $this->state = $this->state->moveForward();
    }
}

$car = new Car;
$car->moveForward();
$car = new Car;
$car->startEngine();
$car->moveForward();

/*
Have to start engine first!
Started Engine.
Moved Car forward...
*/
